src/snowflake/snowpark/functions.py

```diff
@@ -8700,3 +8701,123 @@ def make_interval(
+        >>> new_df.show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        ---------
+        <BLANKLINE>
+
+    Example 2::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()])
+        >>> new_df.show()
+        -----------------
+        |"C_1"  |"C_2"  |
+        -----------------
+        |a      |30     |
+        -----------------
+        <BLANKLINE>
+
+    Example 3::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()], output_column_names=['col1', 'col2'])
+        >>> new_df.show()
+        -------------------
+        |"COL1"  |"COL2"  |
+        -------------------
+        |a       |30      |
+        -------------------
+        <BLANKLINE>
+
+    Example 4::
+        >>> df = session.create_dataframe([10, 20, 30])
+        >>> new_df = map(df, lambda x: x * x, output_types=[IntegerType()], wrap_row=False)
+        >>> new_df.sort(new_df[0]).show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        |400    |
+        |900    |
+        ---------
+        <BLANKLINE>
+
+    Note:
+        The result of the `func` function must be either a scalar value or a tuple containing the same number of elements
+        as specified in the `output_types` argument.
+    """
+    if "packages" not in kwargs:
+        kwargs["packages"] = ["snowflake-snowpark-python"]
+    if len(output_types) == 0:
+        raise ValueError("output_types cannot be empty.")
+    input_types = [field.datatype for field in dataframe.schema.fields]
+    num_fields = len(dataframe.schema.fields)
+    output_schema = [
+        StructField(f"c_{i + 1}", output_types[i]) for i in range(len(output_types))
+    ]
+    output_cols = [
+        col(f"${i + 1}") for i in range(num_fields, num_fields + len(output_types))
+    ]
+    input_row_obj = snowflake.snowpark.Row(*dataframe.columns) if wrap_row else None
+
+    def wrap_result(result):
+        if isinstance(result, list) or isinstance(result, snowflake.snowpark.Row):
+            return tuple(result)
+        elif isinstance(result, tuple):
+            return result
+        else:
+            return (result,)
+
+    class _MapFunc:
+        def process(self, *argv):
+            if wrap_row:
+                argv = [input_row_obj(*argv)]
+            result = func(*argv)
+            yield wrap_result(result)
+
+    map_udtf = dataframe._session.udtf.register(
+        _MapFunc,
+        output_schema=StructType(output_schema),
+        input_types=input_types,
+        is_permanent=False,
+        **kwargs,
+    )
+
+    if not output_column_names:
+        output_column_names = [field.name for field in output_schema]
+
+    _output_cols = [
+        column.alias(desired_name)
+        for column, desired_name in zip(output_cols, output_column_names)
+    ]
```

Comment:
why not integrate this when defining the schema

Reply:
We can set the names when creating the schema, however we still need this code and the last `.select` projection since `join_table_function` includes all the columns from the original DataFrame. 

For example:

```python
another_df = session.create_dataframe([('a',10), ('b', 20), ('d', 30)], 
                                      schema=StructType([StructField('name', StringType()),
                                                         StructField('numval', IntegerType())]))

r1 = map(another_df, lambda row: (row[0] + "_", row[1] * row[1]), [StringType(), IntegerType()], output_column_names=['A','B'])
r1.show()
```
This generates:
```
---------------------------------
|"NAME"  |"NUMVAL"  |"A"  |"B"  |
---------------------------------
|a       |10        |a_   |100  |
|b       |20        |b_   |400  |
|d       |30        |d_   |900  |
---------------------------------
```

Notice that in this case it is assumed that the resulting DataFrame should have two columns.  By including the original columns we can introduce bugs for code that uses column access by index . For example:
```
r1 = map(another_df, lambda row: (row[0] + "_", row[1] * row[1]), [StringType(), IntegerType()], 
r2 = map(r1, lambda r: r[1] * 2, [IntegerType()])
```